spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -315,6 +338,99 @@ contract RolloverVault is
+        // 4) We check if the given bond is NOT malicious, by examining if at least one of the bond's tranches
+        //    is already in the vault from previous deployments. And if the parent bond of the tranche already in the vault
+        //    is the given bond.
+        if (
+            bond.collateralToken() != address(underlying) ||
+            bt.tranches.length <= 1 ||
+            bond.isMature() ||
+            !_isDeployedBond(bond, bt)
+        ) {
+            revert InvalidBond(bond);
+        }
+
+        // track the vault's TVL before the match operation
+        uint256 tvlBefore = getTVL();
+
+        // calculate total amount to be used by summing available amount from the user and the vault's balance
+        uint256[] memory trancheAmtsUsed = new uint256[](bt.tranches.length);
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            trancheAmtsUsed[i] = trancheAmtsIn[i] + bt.tranches[i].balanceOf(address(this));
+        }
+
+        // compute proportional tranche amounts which can be redeemed for the underlying collateral
+        trancheAmtsUsed = bt.computeRedeemableTrancheAmounts(trancheAmtsUsed);
+
+        // computes the collateralizations for each tranche in the current immature bond
+        uint256[] memory trancheUnderlyingBalances;
+        uint256[] memory trancheSupplies;
+        (trancheUnderlyingBalances, trancheSupplies) = bond.getImmatureTrancheCollateralizations(bt);
+
+        // the user's share of the underlying collateral to be returned
+        uint256 underlyingAmt = 0;
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            // compute the user's tranche balance to be user
+            // transfer tranche balance out of the user's wallet to the vault
+            trancheAmtsIn[i] = MathUpgradeable.min(trancheAmtsIn[i], trancheAmtsUsed[i]);
+            IERC20Upgradeable(bt.tranches[i]).safeTransferFrom(_msgSender(), address(this), trancheAmtsIn[i]);
+
+            // pre-calculate the user's share
+            underlyingAmt += trancheUnderlyingBalances[i].mulDiv(trancheAmtsIn[i], trancheSupplies[i]);
+        }
+
+        // ensure that the user is sending in tranches with some redeemable value.
+        if (underlyingAmt <= 0) {
+            revert ValuelessAssets();
+        }
+
+        // redeem the underlying using pooled balances
+        bond.redeem(trancheAmtsUsed);
+
+        // deduct fee before returning to the user
+        underlyingAmt -= _computeMeldFee(bond, underlyingAmt);
+
+        // transfer to the user
+        underlying.safeTransfer(_msgSender(), underlyingAmt);
+
+        // sync vault's tranche balances
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            _syncAndRemoveDeployedAsset(bt.tranches[i]);
+        }
+
+        // sync underlying
+        _syncAsset(underlying);
+
+        // assert that the vault's TVL does not decrease after this operation
+        if (getTVL() < tvlBefore) {
```

Comment:
Rounding errors, if any sloppy math exists in the codebase, could trip this alarm as a false positive. But I do like keeping this strict check as much as we're able, since it enforces that MUST round down on anything that leaves the vault.

Reply:
With the withheld fees the vault tvl in almost all cases should be strictly higher (even with the rounding down)